Added -Xi to include new fields in JSON output

marler8997 opened the pull request:
Adds a new option (hidden for now) that allows an application to select what it wants in the JSON output, i.e.
```
dmd -Xi=compilerInfo -Xi=semantics
```

This would produce a JSON file looking like this:
```
{
    "compilerInfo" : {
        ...
    },
    "semantics": {
        ...
    }
}
```

The option is hidden because the only users of the option (currently known) will be `rdmd` and `dub`.

NOTE: the `-probe` equivalent would be:
```
dmd -Xf=- -Xi=compilerInfo
```
** json output now works even with no input files

marler8997 commented:
NOTE: I also plan on adding the extra fields that were added in (https://github.com/dlang/dmd/pull/7521), however, I thought that should be it's own set of changes.

Hopefully we can get this one merged quickly so I can add the new fields from `-probe` with a good amount of time before the next release.

wilzbach commented on src/dmd/json.d:
```diff
@@ -857,17 +858,165 @@ public:
+    }
+
+    /**
+    Returns true if `c` is a valid character for the start of a name..
```
double `.`

wilzbach reviewed (approved):
Nice. That looks _a lot_ better than your previous approach. 
I'm fully in favor of this PR and hope we can get it as soon as possible ;-)

timotheecour commented on src/dmd/json.d:
```diff
@@ -857,17 +858,165 @@ public:
-    json.arrayEnd();
-    json.removeComma();
+
+    if (global.params.jsonQuery is null)
```
```
if global.params.jsonQuery is null
  global.params.jsonQuery = "modules"; // or whatever we want default query to be; maybe put all the queries here

# then the rest can assume we always have `jsonQuery`
```

marler8997 replied:
Its definitely better to not break backwards compatibility.  If at some point in the future someone determines it's OK in this case, then it's simple to remove the special case.

timotheecour commented on src/dmd/json.d:
```diff
@@ -857,17 +858,165 @@ public:
+
+        json.objectStart();
+        auto parser = QueryParser(global.params.jsonQuery);
+        for (;;)
```
`while(true)` seems more idiomatic?

wilzbach replied:
FYI: `for (;;)` is the "idiomatic" way of Phobos. We should probably put such trivias in the style guide, s.t. they don't constantly pop up...

timotheecour commented on src/dmd/json.d:
```diff
@@ -857,17 +858,165 @@ public:
+            {
+                //json.generateError(format("invalid query: expected name but got '%s' (0x%02x)",
+                //    *parser.next, cast(ubyte)*parser.next));
+                json.generateError("invalid query: expected name or EOF but got something else".ptr);
```
most useful diagnostic would be a to show byte offset from start of string where parser got confused

wilzbach replied:
(also this is uncovered)

marler8997 replied:
workin on this one

marler8997 replied:
ok, I ended up using vsnprintf, also added tests.

wilzbach commented on src/dmd/mars.d:
```diff
@@ -360,6 +360,11 @@ private int tryMain(size_t argc, const(char)** argv)
+        if (global.params.jsonQuery)
+        {
+            generateJson(null);
+            return EXIT_SUCCESS;
```
Ah nice. I didn't see this initially. 

This is more or less what I am trying to achieve in https://github.com/dlang/dmd/pull/7840

wilzbach commented on src/dmd/mars.d:
```diff
@@ -998,6 +970,45 @@ private int tryMain(size_t argc, const(char)** argv)
+    {
+        // Write to stdout; assume it succeeds
+        size_t n = fwrite(buf.data, 1, buf.offset, stdout);
+        assert(n == buf.offset); // keep gcc happy about return values
```
Needs a test.
Example:

```sh
#!/usr/bin/env bash

set -ueo pipefail

name="$(basename "$0" .sh)"
dir="${RESULTS_DIR}/compilable/"
out="$dir/${name}.json.out"

"$DMD" -X | ${RESULTS_DIR}/sanitize_json > "$out"
diff "$out" compilable/extra-files/$name.json
rm "$out"

"$DMD" -Xf=- | ${RESULTS_DIR}/sanitize_json > "$out"
diff "$out" compilable/extra-files/$name.json
rm "$out"
```

(`sanitize_json` currently doesn't accept stdin)

timotheecour replied:
diff can take `-`: https://stackoverflow.com/a/9847527/1426932

```
"$DMD" -X | ${RESULTS_DIR}/sanitize_json  | diff - compilable/extra-files/$name.json
``

marler8997 replied:
> This is existing code, I think any changes to it would be a seperate PR.
ditto

wilzbach commented on src/dmd/mars.d:
```diff
@@ -1869,6 +1880,11 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
                     break;
+                case 'q':
+                    if (!p[3])
+                        goto Lnoarg;
```
test

marler8997 replied:
done

timotheecour commented on src/dmd/mars.d:
```diff
@@ -998,6 +970,45 @@ private int tryMain(size_t argc, const(char)** argv)
+
+    // Write buf to file
+    const(char)* name = global.params.jsonfilename;
+    if (name && name[0] == '-' && name[1] == 0)
```
`strcmp`

marler8997 replied:
This is existing code, I think any changes to it would be a seperate PR.

timotheecour commented on src/dmd/json.d:
```diff
@@ -857,17 +858,165 @@ public:
+        // of modules representing their syntax.
+
+        json.arrayStart();
+        for (size_t i = 0; i < modules.dim; i++)
```
foreach

timotheecour commented on src/dmd/json.d:
```diff
@@ -857,17 +858,165 @@ public:
+            auto fieldID = tryParseJsonField(fieldName);
+            if (!fieldID.hasValue)
+            {
+                json.generateError(("invalid query: unknown field name '" ~ fieldName ~ "'\0").ptr);
```
aren't backticks "`" used instead of "'"  in messages ?

timotheecour commented on src/dmd/json.d:
```diff
@@ -857,17 +858,165 @@ public:
+            auto fieldID = tryParseJsonField(fieldName);
+            if (!fieldID.hasValue)
+            {
+                json.generateError(("invalid query: unknown field name '" ~ fieldName ~ "'\0").ptr);
```
`.ptr` is not needed; D string litterals are null terminated and can decay to `const(char)*`
+everywhere relevant

marler8997 replied:
Doesn't appear to work in this case

timotheecour replied:
looks like it did :) (not seeing .ptr after string literals in your latest commit) 
if not, i'm curious in what case it didn't work? i thought string literals would convert?

marler8997 replied:
The latest doesn't using `~` to append any strings.  Looks like appending strings breaks implicit conversion from `string` to `const(char)*`.

timotheecour commented on src/dmd/json.d:
```diff
@@ -857,17 +858,165 @@ public:
+    JsonField value;
+}
+OptionalJsonField tryParseJsonField(const(char)[] fieldName)
+{
```
DRY:
```
foreach(a; __traits(allMembers, JsonField))
  if(fieldName == a) return OptionalJsonField(true, mixin(`JsonField.`~a));
```

marler8997 replied:
done, I was thinking I needed to use phobos `EnumMembers`...I forgot I could do this with `__traits`.

timotheecour commented on src/dmd/mars.d:
```diff
@@ -998,6 +970,45 @@ private int tryMain(size_t argc, const(char)** argv)
+        else
+        {
+            // Generate json file name from first obj name
+            const(char)* n = global.params.objfiles[0];
```
`assert(global.params.objfiles.length > 0);` ?

marler8997 replied:
> This is existing code, I think any changes to it would be a seperate PR.

ditto

timotheecour commented on src/dmd/mars.d:
```diff
@@ -998,6 +970,45 @@ private int tryMain(size_t argc, const(char)** argv)
+    }
+    else
+    {
+        /* The filename generation code here should be harmonized with Module::setOutfile()
```
yup!

timotheecour commented on test/compilable/extra-files/json2.out:
```diff
@@ -0,0 +1,62 @@
+        "cwd": "VALUE_REMOVED_FOR_TEST",
+        "importPaths": [
+            "compilable",
+            "..\/..\/druntime\/import",
```
how about making sanitize replace `\/` with `/` ?

marler8997 replied:
This is the JSON generated by `std.json` (not dmd).  Dmd doesn't generate these weird escaped slashes.  To fix this we would need to post-process the JSON after it is sanitized, basically another JSON parser that sanitizes after `std.json`...lol :)

timotheecour replied:
lol indeed... std.json really sux

marler8997 replied:
wait, there already is an option for this! https://dlang.org/phobos/std_json.html#.JSONOptions.doNotEscapeSlashes

I'll updated the PR to use it :)

timotheecour commented on test/sanitize_json.d:
```diff
@@ -39,25 +39,61 @@ int main(string[] args)
+        foreach (ref obj; root.array)
+        {
+            auto kind = obj.object["kind"].str;
+            if (kind == "compilerInfo")
```
DRY + type safety:
s/"compilerInfo"/JsonField.compilerInfo.stringof/
+ everywhere relevant

marler8997 replied:
This is the test suite, not dmd.  It doesn't have access to the `JsonField` type defined in the compiler.

timotheecour replied:
`import dmd.json : JsonField;` ?

wilzbach replied:
Yeah, the testsuite shouldn't (directly) depend on what it is testing.

timotheecour commented on src/dmd/mars.d:
```diff
@@ -1869,6 +1880,11 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
+                case 'q':
+                    if (!p[3])
+                        goto Lnoarg;
+                    params.jsonQuery = p + 3 + (p[3] == '=');
```
check p[3] == '=' and `params.jsonQuery = p + 4`
We really should not allow things like `-Xqfoo` in new flags; this is old baggage from dmd cmd line (eg `-offile` instead of (`-of=file`), the downsides have these have been discussed: it's not standard, hard to read visually, and disallows future new flags (eg `-Xquery=baz`)

marler8997 replied:
Yes I agree with this, I'll remove support for omitting `=`.

timotheecour commented:
* I suggest making flag `-Xf-` documented instead of undocumented since it's used externally (rdmd and especially dub are external to dmd)
* toplevel commit msg should mention new behaviors:
** `-Xf=-` means stdout
** json output now works even with no input files

wilzbach commented:
> toplevel commit msg should mention new behaviors:

What do you mean with "top-level"  - wouldn't be the merge commit ?
A (short) changelog entry is much more valuable to the user.

marler8997 commented:
Ok I changed `-Xf-` to `-Xf=-`.  However, `-Xf=-` was existing functionality not introduced by this PR.

timotheecour commented on src/dmd/json.d:
```diff
@@ -857,17 +880,161 @@ public:
+            auto fieldName = parser.tryParseName();
+            if (fieldName is null)
+            {
+                json.generateError("invalid query at offset %d: expected name but got '%c' (0x%02x)",
```
```
'%c' => `%c`
```

marler8997 replied:
I think the single-quotes makes more sense when printing a raw character.

wilzbach replied:
Just "name" isn't so helpful, how about generating an up-to-date list?

```d
enum names = () {
    string s;
    static immutable delim = ", ";
    foreach (member; __traits(allMembers, JsonField))
        s ~= member.stringof ~ delim;
    return s[0 .. $ - delim.length];
}();
```

marler8997 replied:
The first error is caused by an invalid character, providing all the field names might be too much information in this case.  However, providing the names for the the 2nd error, which is an invalid field name definitely makes sense.  I've included this change for the 2nd error.

timotheecour replied:
> I think the single-quotes makes more sense when printing a raw character.

then how about doing like:
src/dmd/parse.d:7425:77:                                error("mismatched string literal postfixes `'%c'` and `'%c'`", postfix, token.postfix);

 the point is it'll get syntax highlighted

wilzbach commented on changelog/json_query.dd:
```diff
@@ -0,0 +1,23 @@
+Added the `-Xq=...` option to support limiting JSON output to a set of fields.
+
+Using the `-Xq=...` option limits the JSON output to a set of fields, i.e.
+---
```
The `CONSOLE` macro is specifically made for this:

```
$(CONSOLE dmd -Xq=compilerInfo,semantics)
```

see e.g. https://dlang.org/changelog/pending.html#deprecation-dylib

wilzbach reviewed (approved):
Nice! I only found a few more nits, but it looks more than ready to go to me.
I really like the extra effort you made with adding all these additional test files :+1:

marler8997 commented:
I've thought about it and I'm going to make a small change.  I'm going to check the JSON query during command line parse time instead of JSON generation time.  I wasn't sure which would be better but for some reason I've changed my mind...

timotheecour commented:
agreed, we need to fail early

marler8997 commented:
Hmmm, getting a bunch of failures with this message:
```
test_results/d_do_test: symbol lookup error: test_results/d_do_test: undefined symbol: _D3std9exception__T7enforceZ__TQmTbZQrFNaNfbLAxaAyamZb
```
Looks like some of the CI builders aren't able to build the `d_do_test` tool because of a linker error...I wasn't able to reproduce this in my box...

wilzbach commented:
Uhm I have seen this one on https://github.com/dlang/dmd/pull/7845 too and I suspect it has something to do with the recently merged https://github.com/dlang/phobos/pull/6086

wilzbach commented:
FYI I'm proposing to revert the Phobos PR for now (-> revert: https://github.com/dlang/phobos/pull/6130), s.t. we have time to figure it out and check why the CIs didn't catch it.

marler8997 commented:
Ok, sounds like a good idea.  We're still getting that OOM issue on linux intermittently...wondering how we can address that one.  The autotester has 2GB of memory according to @braddr, and I think that should be enough without intermittently failing.  Maybe we can wrap the processes that are running the tests in phobos, and on OOM errors retry the test...? Not sure.

wilzbach commented:
>  We're still getting that OOM issue on linux intermittently...wondering how we can address that one.

Repeating what I did here: https://github.com/dlang/phobos/pull/6061 or fix the CTFE parser :P 
`std/regex/internal/tests.d` is still a really, really bad guy in terms of regex performance...

marler8997 commented:
Ah ok, sounds like a plan.

marler8997 commented:
@wilzbach looks like you fixed it :)

wilzbach commented:
BTW @marler8997 I added support for `-Xf=-` to run.dlang.io 

https://run.dlang.io/is/mA9krz - you don't even need to enter it. There's a button.

![image](https://user-images.githubusercontent.com/4370550/35881738-188cc6ea-0b82-11e8-86a0-e82fe4acaf31.png)

marler8997 commented:
Let's get this one merged soon.  After this one I'll create a new PR that adds the rest of the fields from the originally proposed `-probe` flag and finalize the field names.  After that PR is done that should be everything to allow me to finish the `rdmd` change for single invocation.  If we get this PR and the next one done soon we can make it before the next release.

wilzbach commented:
> Let's get this one merged soon.

I just waiting for someone like @WalterBright or @andralex to have a brief look and approve this direction. I doubt that Walter cares much about the Json output (as long as it doesn't affect codegen) nor break anything.
If no one raises any voice until this Friday evening, I will go ahead and merge this.

> After this one I'll create a new PR that adds the rest of the fields from the originally proposed -probe flag and finalize the field names. After that PR is done that should be everything to allow me to finish the rdmd change for single invocation. If we get this PR and the next one done soon we can make it before the next release.

:+1:

WalterBright commented on src/dmd/globals.d:
```diff
@@ -182,6 +182,7 @@ struct Param
 
     bool doJsonGeneration;              // write JSON file
     const(char)* jsonfilename;          // write JSON file to jsonfilename
+    uint jsonQueryFlags;                // flags for JSON query
```
Shouldn't this be typed as `JsonQueryFlags` ?

WalterBright commented on src/dmd/json.d:
```diff
@@ -857,17 +871,178 @@ public:
+
+/**
+Parses a JSON query into a set of flags.
+
```
`Params:` ?

WalterBright commented on src/dmd/json.d:
```diff
@@ -857,17 +871,178 @@ public:
+
+Returns: false on error, prints its own error messages.
+*/
+bool parseJsonQuery(const(char)* query, uint* outQueryFlags)
```
`out uint outQueryFlags`

WalterBright replied:
or `ref` as it is cumulative.

marler8997 replied:
Ugh, I hate that now the caller won't know their "value" argument could be modified, i.e.
```D
ubyte myFlags = 0;
parseJsonQuery(myQuery, myFlags);
```
Now `myFlags` was modified, but it's not clear from the call that this happened.  I would prefer to see:
```D
ubyte myFlags = 0;
parseJsonQuery(myQuery, &myFlags);
```
Now you definitely know that `myFlags` could have been modified.  That's just my preference, but this is not my code so I'll do whatever the core team prefers for `dmd`.

WalterBright commented on src/dmd/mars.d:
```diff
@@ -1869,6 +1885,16 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
+                        goto Lerror;
+                    if (!p[4])
+                        goto Lnoarg;
+                    if (!parseJsonQuery(p + 4, &global.params.jsonQueryFlags))
```
I see this is cumulative to jsonQueryFlags. Should it be? There's no way to turn off a flag with them being cumulative.

marler8997 replied:
All flags are off by default, you "turn then off" by not specifying them.  Are you saying that we should have a way to allow a user to turn it on in one argument, and they specify another argument to turn it off?

WalterBright commented on src/dmd/mars.d:
```diff
@@ -1869,6 +1885,16 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
+                    if (!p[4])
+                        goto Lnoarg;
+                    if (!parseJsonQuery(p + 4, &global.params.jsonQueryFlags))
+                        return true; // error
```
should do a `continue` like the rest, not `return`

WalterBright commented on src/dmd/json.d:
```diff
@@ -857,17 +871,178 @@ public:
+*/
+bool parseJsonQuery(const(char)* query, uint* outQueryFlags)
+{
+    auto parser = QueryParser(query);
```
This seems like a lot of effort to simply turn a comma-separated list of 4 identifiers into 4 flags? In fact, this could be avoided entirely by just supporting multiple -Xq=identifier queries, which you already do:
```
dmd -Xq=compilerInfo,buildInfo,semantics,modules
```
becomes:
```
dmd -Xq=compilerInfo -Xq=buildInfo -Xq=semantics -Xq=modules
```
and all this code goes away. Note we do the same thing for multiple `-version=identifier` switches, etc.

timotheecour replied:
there's a precedent for that with:
* `-i=+core,-std.stdio`   
* and in clang, `-Wl,-lfoo,-lbar`
* and the suggested -unittest=std.stdio,foo.bar

maybe exposing that logic (parsing comma separated list) into reusable code would be an option

WalterBright replied:
I would have argued against the -i syntax, on the same grounds. What clang does is not that compelling.

In any case, reinventing another parser for CSV options is not the best plan. As noted in the comments, dmd already has several different ways of specifying a file, which is a miserably stupid design, and is really hard to unwind because of subtle breakage. Let's not do that with multiple CSV parsers with same-but-maddeningly-different behaviors.

marler8997 replied:
If you prefer using multiple arguments, i.e.
```
-i=foo -i=-foo.bar -i=std -i=core
```
 I don't think it's too late to change `-i`.  It hasn't showed up in a major release yet.  I agree that consistency is virtue we should aspire to.  Just say the word and I'll create a PR for this.

marler8997 replied:
I've created a PR to remove the "comma support" for `-i`: https://github.com/dlang/dmd/pull/7857/files

timotheecour commented on src/dmd/json.d:
```diff
@@ -797,19 +797,33 @@ public:
         objectEnd();
     }
 
+    private void generateModules(Modules* modules)
```
even private? 
from https://dlang.org/dstyle.html: 
`All public declarations will be documented in Ddoc format and should have at least Params and Returns sections.` => says public

WalterBright replied:
If they're non-trivial, yes. I don't know what `generateModules()` does. How can it generate a module when it is passed modules? Where do these generated modules go? I'm sure this is obvious to you since you wrote it, but I submit if you looked at it a year from now, you wouldn't know what it did. (This happens to me all the time.)

timotheecour replied:
I'm not questioning whether it should have a comment, I'm questioning whether it should have a comment formatted as Ddoc

marler8997 commented:
Ok changes made.

marler8997 commented:
Note, now that we moved from one query (`-Xq=<query>`) to multiple fields (`-Xi=<field1> -Xi=<field2>`), I've changed the wording from "query" to "include field", hence the change from `-Xq` to `-Xi`.

wilzbach commented on src/dmd/json.d:
```diff
@@ -857,17 +891,77 @@ public:
+    string prefix = "";
+    foreach (enumName; __traits(allMembers, JsonFieldFlags))
+    {
+        if (enumName != "none")
```
How about:

```d
foreach (idx, enumName; __traits(allMembers, JsonFieldFlags))
{
    if (idx > 0)
    {
        s ~= ", " ~ enumName;
    }
}
```

marler8997 replied:
Hmmm, that produces a different string ", compilerInfo, ..." instead of "compilerInfo, ..."

wilzbach replied:
Ah it's too late here. I meant:

```d
if (idx > 0)
    s ~= ", ";
s ~= enumName;
```

(the main point was about not hard-coding that the first `enumName` is named "none")

marler8997 replied:
Oh...but now you're hardcoding that the first member should be ignored...is that better than hardcoding that the value named "none" should be ignored?

marler8997 replied:
Eh...I suppose I don't have a preference for either...so we'll go with your preference :)

wilzbach commented on src/dmd/json.d:
```diff
@@ -857,17 +891,77 @@ public:
+    auto fieldNameString = fieldName[0 .. strlen(fieldName)];
+    foreach (enumName; __traits(allMembers, JsonFieldFlags))
+    {
+        if (enumName != "none" && fieldNameString == enumName)
```
You could move `if (enumName != "none")` out of the loop.
That saves a few duplicated comparisons.

wilzbach commented on test/sanitize_json.d:
```diff
@@ -39,25 +39,50 @@ int main(string[] args)
+            {
+                sanitizeSemantics(semantics.object);
+            }
+        }
```
DRY idea: 

```d
static foreach (e; [tuple("compilerInfo", "sanitizeCompilerInfo")
                    tuple("buildInfo", "sanitizeBuildInfo")])])
{{
    auto infonfo = rootObject.get(e[0], JSONValue.init);
    if (compilerInfo.type != JSON_TYPE.NULL)
        mixin(e[1] ~ "(info.object);");
}}
```

It's okay to use newer compiler features here as DMD + DRuntime + Phobos are just freshly built from master.

wilzbach commented on test/sanitize_json.d:
```diff
@@ -39,25 +39,44 @@ int main(string[] args)
+    if (s.length == 0 || (s[0] < 'a' || s[0] > 'z'))
+        return s;
+    return (cast(char)(s[0] - ('a' - 'A'))) ~ s[1 .. $];
+}
```
https://dlang.org/phobos/std_string.html#.capitalize ?

marler8997 replied:
Looks like that one won't work, it converts the rest to lower-case...which is not what is wanted.

wilzbach replied:
Ah that's unexpected. I thought there's a function for this in Phobos :/
Well, you could do this:

```d
import std. unit : asCapitalized;
return s.take(1).asCapitalized.chain(s.drop(1));
```

It would be more idiomatic, but put more stress on the CTFE (though it's a small standalone script for which this is probably not a concern);

marler8997 replied:
Ugh...I can't get it to work.  I added `.array` to the end but it ends up returning `dchar[]`...

wilzbach replied:
Interesting it worked for me (https://run.dlang.io/is/xP19Bv), 

>  the end but it ends up returning dchar[]

You can use `.byCodeUnit` to opt-out of auto-decoding, but `to!string` or `text` from `std.conv` are both smart enough (e.g. https://run.dlang.io/is/zv8RlX)

For pure ASCII, there's also `std.ascii.toLower`:

```d
s.enumerate.map!(a => a.index == 0 ? a.value.toUpper : a.value).text;
```

https://run.dlang.io/is/rHnADh

But this is definitely nit-picking here ;-)

marler8997 replied:
looks like `std.conv.text` did the trick.  My `std.range` / `std.algorithm` 'fu could use work.

wilzbach reviewed (approved):
I still like it :+1:

JinShil commented:
This PR could use a statement regarding motivation (at least if you need my review).  I don't have an intimate understanding of the problem this solves or the use case.  Also, why do we need multiple options?  Is there a problem with just generating everything?

Also, I don't see why it shouldn't be documented.  What's the motivation for keeping it hidden?  Are we not sure it's a good idea or not?

The code is no 100% covered by tests.  Can the lack of coverage be justified?

Overall, though, I don't have any issue with this PR, and I'm not objecting to it;  I'm actually trying to move it forward by adding my support.  But I need some understanding (answers to the questions above) before I can support it.

marler8997 commented:
> This PR could use a statement regarding motivation (at least if you need my review)

This PR is a result of a suggestion from @WalterBright .  We needed a way of capturing information from the compiler without redirecting stdout/stderr.  I suggested providing a way to redirect verbose output to a file but Walter suggested we add the information to the JSON file, so this PR is the result.

> Also, why do we need multiple options? Is there a problem with just generating everything?

Generating everything is fine, but the output is meant for tools that can process the JSON output.  So it makes sense to allow the tool to specify what it wants.  This also has the advantage that the tool won't need to pay the overhead to generate everything if they don't need it.  Also it provides a mechanism to allow a tool to run the compiler with no source files, something DUB will do to query the compiler information.  It also maintains backwards compatibility. Existing use cases where JSON is generated is unnafected, it generates the original format, but if a tool uses `-Xi` then it will generate the new format with the requested fields.  And it will remain forward compatible because any new fields that might be added later will not be included since they won't be requested via `-Xi`.

> Also, I don't see why it shouldn't be documented. What's the motivation for keeping it hidden? Are we not sure if it's a good idea?

The only reason for not documenting it is no one has thought of a good use case for this feature to be used out in the wild.  The only current use cases are `dub` and `rdmd`, both of which are internal tools.  If it deemed to be useful outside of internal use, then documenting it makes sense.  I'm basically taking the conservative approach where I'm not including any changes that can always be decided later.

> The code is not 100% covered by tests. Can the lack of coverage be justified?

I haven't checked the code coverage. Depends on which code paths aren't taken, I'll see if I can take a look.

JinShil commented:
Thanks @marler8997 

Who are the primary consumers of this JSON output and for what purpose?  You mentioned `dub` and `rdmd`, so what do they need with this information?

What about feature overlap?  Isn't this information something 3rd party tools can obtain for themselves using DMD-as-a-library?

marler8997 commented:
rdmd will use it to output all the files used during a compilation.  It uses this for dependency analysis. It will also use compilerInfo to know what features a particular compiler binary supports. Same with dub.

Dmd as a library really isn't applicable for these use cases. The JSON output is used to get information about a particular compiler binary or a particular build, whereas dmd as a library shows an application to have access to compiler functionality.

JinShil commented on src/dmd/mars.d:
```diff
@@ -998,6 +970,50 @@ private int tryMain(size_t argc, const(char)** argv)
+        {
+            if (global.params.objfiles.dim == 0)
+            {
+                error(Loc.initial, "cannot determine JSON filename, use -Xf=<file> or provide a source file");
```
Tick marks:  e.g. \`-Xf=<file>\`

JinShil commented on test/fail_compilation/jsonBadField1.d:
```diff
@@ -0,0 +1,9 @@
+REQUIRED_ARGS: -Xi=UNKNOWN_FIELD_NAME
+TEST_OUTPUT:
+---
+Error: unknown JSON field `-Xi=UNKNOWN_FIELD_NAME`, expected one of compilerInfo, buildInfo, modules, semantics
```
Please modify the code to surround each expected option in tick marks, e.g. \`compilerInfo\`, \`buildInfo\`, \`modules\`, \`semantics\`

marler8997 replied:
done, thanks for the suggestion

Commit:
Added -Xi to include new fields in JSON output

Fix issue 18367

wilzbach commented:
> This also has the advantage that the tool won't need to pay the overhead to generate everything if they don't need it. For example, dub only needs to generate compilerInfo and generating it quickly will likely be a requirement. 

Yes, at the moment Dub creates and compiles a probing file and invocates DMD on it to parse the `pragma(msg)` output  _every time_ before it builds something  (see https://github.com/dlang/dmd/pull/7521).

> For example, 

There are many other potential use cases - run.dlang.io is another or basically any build tool that needs more information about the currently installed D compiler.

> Also, I don't see why it shouldn't be documented. What's the motivation for keeping it hidden? Are we not sure if it's a good idea?

Yep this was done, s.t. we have one or two releases to experiment with until we set it into stone.

JinShil commented:
Thank you @marler8997 and @wilzbach.  I see where this is going now.  I'm just waiting on coverage.
